services/src/main/java/org/keycloak/authentication/ClientAuthenticationFlow.java

```diff
@@ -69,6 +72,18 @@ public Response processFlow() {
             logger.debugv("client authenticator: {0}", factory.getId());
 
             AuthenticationProcessor.Result context = processor.createClientAuthenticatorContext(model, authenticator, executions);
+
+            String clientAuthenticatorType = defaultClientAuthenticatorType;
+            if (context.getClient() != null && context.getClient().getClientAuthenticatorType() != null) {
```

Comment:
It looks to me that `context.getClient()` will be non-null just if some authenticator managed to authenticate client, which is different than the authenticator set on the client. I can see some possible security issues if we rely on the client, which was set here by the incorrect authenticator than the correct authenticator type for target client...

I wonder if this is also the reason of the test failures of this PR?

I see alternative approach might be to make sure that the check for "token reuse" is performed just once per-request as I've suggested in this comment https://github.com/keycloak/keycloak/issues/34547#issuecomment-2454687261 . WDYT?